### ISSUE TITLE
cabal doesn't install on ghc 8.8.1

### ISSUE THREAD

**Issue by GeorgeCo** [haskell/directory#98]
cabal --version
cabal --version
cabal-install version 3.0.0.0
compiled using version 3.0.0.0 of the Cabal library 
bash-3.2$ cabal install directory
cabal install directory
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: directory-1.3.4.0 (user goal)
[__1] next goal: ghc (user goal)
[__1] rejecting: ghc-8.8.1/installed-8.8... (conflict: directory==1.3.4.0, ghc
=> directory==1.3.3.2/installed-1.3...)
[__1] rejecting: ghc-8.6.5, ghc-8.6.4, ghc-8.6.1, ghc-8.4.4, ghc-8.4.3,
ghc-8.4.1, ghc-8.2.2, ghc-8.2.1 (constraint from user target requires ==8.8.1)
[__1] fail (backjumping, conflict set: directory, ghc)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: directory, ghc

**Comment by Rufflewind** [comment]
I believe this caused by the difference in behavior between `cabal v1-install` and `cabal v2-install`.  You are currently using [`v2-install`](https://www.haskell.org/cabal/users-guide/nix-local-build.html#cabal-v2-install), which installs to a [package environment](https://downloads.haskell.org/~ghc/8.8.1/docs/html/users_guide/packages.html#package-environments) file such as `~/.ghc/${arch}-${os}/environments/default`.  In that environment file (which you can inspect with a text editor), `directory-1.3.3.2` is probably already installed, so you cannot install a different version.

To work around this, you can try to install to a new environment, say `./foobar.env`:

    cabal v2-install --lib --package-env=./foobar.env directory-1.3.4.0

**Comment by GeorgeCo** [comment]
Is there a user guide for cabal 3.0? I find it a big change where much of what I used to do doesn't work. These differences feel like regressions to me. I'm concerned that many unsophisticated users will feel the same. Previously we didn't really have to know about package env files. I think new ghc users will find the ecosystem more daunting than previously. I just wanted to give my feedback.

**Comment by cartazio** [comment]
Yeah.  There seems to be a bunch of stuff dropped the ball ux default wise.

**Comment by Rufflewind** [comment]
There is some bit of documentation in the Cabal user guide on the new changes but I do agree it is quite confusing and the migration path is not adequately explained. It did take me some time to figure out what was causing the issue your described.

I do suggest posting your feedback on the Cabal project itself, not here, as I'm not sure whether the maintainers of Cabal would see this thread.

**Comment by GeorgeCo** [comment]
Thanks Rufflewind, cartazio. Sorry it took me so long to post to the Cabal project but I finally did: https://github.com/haskell/cabal/issues/6342

**Comment by GeorgeCo** [comment]
By the way, I was surprised and please to find that " cabal v1-install directory" works fine

**Comment by Rufflewind** [comment]
@GeorgeCo Did you try the workaround suggested above?

    cabal v2-install --lib --package-env=./foobar.env directory-1.3.4.0

**Comment by GeorgeCo** [comment]
I think I did and I think it worked but I'm not anxious to learn about the v2 version of stuff until I understand the motivation for the incompatible changes. I haven't looked hard but I'm unaware of a design doc for cabal 3.0 that explains that. I did check the GHC 8.8.1 User's Guide for a link to some documentation to Cabal 3.0 but I didn't see any. Hopefully I'll learn something from my cabal bug report.
